plan/planbuilder.go

```diff
@@ -951,154 +935,251 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
-			return nil
+	totalTableCols := insertPlan.Table.Cols()
+	for i, valuesItem := range insert.Lists {
+		// The length of the value_list should keep unchanged.
+		// "insert into t values (), ()" is valid.
+		// "insert into t values (), (1)" is not valid.
+		// "insert into t values (1), ()" is not valid.
+		// "insert into t values (1,2), (1)" is not valid.
+		if i > 0 && len(insert.Lists[i-1]) != len(valuesItem) {
+			b.err = ErrWrongValueCountOnRow.GenByArgs(i + 1)
+			return
 		}
-
-		insertPlan.Schema4OnDuplicate = expression.MergeSchema(insertPlan.tableSchema, insertPlan.SelectPlan.Schema())
+		exprList := make([]expression.Expression, 0, len(valuesItem))
+		for j, valueItem := range valuesItem {
+			var expr expression.Expression
+			var err error
+			if dft, ok := valueItem.(*ast.DefaultExpr); ok {
+				if dft.Name != nil {
+					expr, err = b.findDefaultValue(totalTableCols, dft.Name)
+				} else {
+					expr, err = b.getDefaultValue(affectedValuesCols[j])
+				}
+			} else if val, ok := valueItem.(*ast.ValueExpr); ok {
+				expr = &expression.Constant{
+					Value:   val.Datum,
+					RetType: &val.Type,
+				}
+			} else {
+				expr, _, err = b.rewriteWithPreprocess(valueItem, mockTablePlan, nil, true, checkRefColumn)
+			}
+			if err != nil {
+				b.err = errors.Trace(err)
+			}
+			exprList = append(exprList, expr)
+		}
+		insertPlan.Lists = append(insertPlan.Lists, exprList)
 	}
+	insertPlan.Schema4OnDuplicate = insertPlan.tableSchema
+}
 
-	mockTablePlan.SetSchema(insertPlan.Schema4OnDuplicate)
-	onDupCols := make(map[string]struct{}, len(insert.OnDuplicate))
-	for _, assign := range insert.OnDuplicate {
-		// Check whether the column to be updated exists in the source table.
-		col, err := insertPlan.tableSchema.FindColumn(assign.Column)
-		if err != nil {
-			b.err = errors.Trace(err)
-			return nil
-		} else if col == nil {
-			b.err = ErrUnknownColumn.GenByArgs(assign.Column.OrigColName(), "field list")
-			return nil
-		}
+func (b *planBuilder) buildSelectPlanOfInsert(insert *ast.InsertStmt, insertPlan *Insert) {
+	affectedValuesCols, err := insertPlan.getAffectCols(insert)
+	if err != nil {
+		b.err = errors.Trace(err)
+		return
+	}
+	selectPlan := b.build(insert.Select)
+	if b.err != nil {
+		return
+	}
 
-		// Check whether the column to be updated is the generated column.
-		column := columnByName[assign.Column.Name.L]
-		if column.IsGenerated() {
-			b.err = ErrBadGeneratedColumn.GenByArgs(assign.Column.Name.O, tableInfo.Name.O)
-			return nil
-		}
+	// Check that the length of select' row is equal to the col list.
+	if selectPlan.Schema().Len() != len(affectedValuesCols) {
+		b.err = ErrWrongValueCountOnRow.GenByArgs(1)
+		return
+	}
 
-		// Construct the function which calculates the assign value of the column.
-		expr, err := b.rewriteInsertOnDuplicateUpdate(assign.Expr, mockTablePlan, insertPlan)
-		if err != nil {
-			b.err = errors.Trace(err)
-			return nil
+	// Check that there's no generated column.
```

Comment:
@CaitinChen @lilin90 PTAL this comment, thanks

Reply:
Check to guarantee that no column is generated.